ext/drstatecmp/drstatecmp.dox

```diff
@@ -0,0 +1,146 @@
+saved. This state will be used for state restoration before the re-execution of the basic
+block. At the end of the original block, code is inserted to save the state for later
+comparison and restore the machine state to the state before executing this bb. This stage
+also removes the previously inserted jump instr to the TERM label and let it fall through
+to the instructions of the copy of the basic block. Then, the copy of the basic block is
+instrumented. At the end of the copy, a clean call is inserted that compares the current
+machine state with the machine state that was saved at the end of the original basic
+block. Finally, all previously-inserted (from other passes) instrumentation instructions
+are removed from the basic block copy, making it instrumentation-free and enabling
+identification of various types of instrumentation clobering.
+
+\subsection sec_drstatecmp_checks_side_effects Basic Blocks with Side Effects
+
+Checking basic blocks with side effects is not yet implemented but this section serves
+as a high-level design overview.
+
+Basic blocks with side-effects cannot be executed twice (at least not without a lot of
+extra complexity to monitor and restore memory state). In this case, the checks are
+inserted throughout the basic block whenever needed taking into account the app
+instructions, existing instrumentation, and the lazy restoration by drreg. As a result,
+incomplete-decoder bugs and lazy-condition-ddreg bugs will not be detected.
+
+Instructions with side effects include instructions that write to memory, interrupts,
+and system calls. To avoid inter-procedural analysis, all function calls are
+conservatively assumed to have side effects.
+
+Some of the side effects could be handled in a way that enables re-execution of more
+basic blocks. For example, system calls could be executed once in the original basic
+block and then in the basic block copy the system call is just emulated by using
+the return value seen in the first execution. Other side effects such as storing to
+memory can be executed only once in cases where the stored value is not loaded
+within the basic block. Memory could also be handled by saving and restoring the
+memory state in addition to the machine state. Naturally, monitoring of the written
+memory addresses is needed to limit the memory state size.
+
+
+\section sec_drstatecmp_instr_state_sharing Instrumentation State Sharing
+
+The instrumentation stages communicate through the user_data parameter. Space for this
+data is allocated in thread-local storage in the app2app stage and deallocated at the
+end of the insert instrumentation stage. The app2app stage specifies in the user_data
+whether a basic block is side-effect free or not. The analysis stage is not used except
+for passing over the user_data to the insert instrumentation stage. In the latter stage,
+the user_data is also used to track when the instrumentation moves from the original
+basic block to its copy.
```

Comment:
Hmm, this is more of an internal implementation detail.  I would expect these docs to focus more on how to use this library: like how to create a test from an existing complex client.  I want to create a version that includes these extra state comparisons for my test.  Do I have to use ifdefs to do two separate builds to avoid the checks in my release version?  Or is there a convenient flag where drstatecmp can be disabled without having to be ifdef-ed out?  If I link with it but place the drstatecmp_init call under a test-only flag in my client: will that do it, or will there still be overhead?

What apps should be executed for the test?  Up to the user presumably.

How are comparison failures reported?  Does the user get to decide whether the process is killed or whether there's just a printout or a callback to some user-controlled action?